x-pack/plugins/monitoring/server/alerts/disk_usage_alert.ts

```diff
@@ -0,0 +1,403 @@
+        state: AlertingDefaults.ALERT_STATE.firing,
+        nodes: firingNodes
+          .map((state) => `${state.nodeName}:${state.diskUsage.toFixed(2)}`)
+          .join(','),
+        count: firingCount,
+        clusterName: cluster.clusterName,
+        action,
+        actionPlain: shortActionText,
+      });
+    } else {
+      const resolvedNodes = (alertStates as AlertDiskUsageState[])
+        .filter((state) => !state.ui.isFiring)
+        .map((state) => `${state.nodeName}:${state.diskUsage.toFixed(2)}`);
+      const resolvedCount = resolvedNodes.length;
+
+      if (resolvedCount > 0) {
+        const internalMessage = i18n.translate(
+          'xpack.monitoring.alerts.diskUsage.resolved.internalMessage',
+          {
+            defaultMessage: `Disk usage alert is resolved for {count} node(s) in cluster: {clusterName}.`,
+            values: {
+              count: resolvedCount,
+              clusterName: cluster.clusterName,
+            },
+          }
+        );
+
+        instance.scheduleActions('default', {
+          internalShortMessage: internalMessage,
+          internalFullMessage: internalMessage,
+          state: AlertingDefaults.ALERT_STATE.resolved,
+          nodes: resolvedNodes.join(','),
+          count: resolvedCount,
+          clusterName: cluster.clusterName,
+        });
+      }
+    }
+  }
+
+  private executeDeltas(
+    services: AlertServices,
+    cluster: AlertCluster,
+    newAlertStates: AlertDiskUsageState[],
+    oldAlertStates: AlertDiskUsageState[]
+  ) {
+    const deltaFiringStates = [];
+    const deltaResolvedStates = [];
+    const deltaInstanceIdPrefix: string = `.monitoring:${this.type}:${
+      cluster.clusterUuid
+    }:${Date.now()}:`;
+
+    for (const newAlertState of newAlertStates) {
+      const relatedOldState = oldAlertStates.find(
+        (oldState) =>
+          oldState.nodeId === newAlertState.nodeId &&
+          oldState.ui.isFiring !== newAlertState.ui.isFiring &&
+          oldState.ui.resolvedMS !== newAlertState.ui.resolvedMS
+      );
+      if (!relatedOldState) {
+        if (newAlertState.ui.isFiring) {
+          deltaFiringStates.push(newAlertState);
+        } else if (newAlertState.ui.resolvedMS) {
+          deltaResolvedStates.push(newAlertState);
+        }
+      }
+    }
+
+    if (deltaFiringStates.length + deltaResolvedStates.length === newAlertStates.length) {
+      /** No delta changes, so we do nothing */
+      return;
+    }
+
+    if (deltaFiringStates.length) {
+      const instance = services.alertInstanceFactory(`${deltaInstanceIdPrefix}:firing`);
+      this.executeActions(instance, { alertStates: deltaFiringStates }, null, cluster);
+    }
+
+    if (deltaResolvedStates.length) {
+      const instance = services.alertInstanceFactory(`${deltaInstanceIdPrefix}:resolved`);
+      this.executeActions(instance, { alertStates: deltaResolvedStates }, null, cluster);
+    }
+  }
+
+  protected async processData(
```

Comment:
What if we did something like this for `processData` to handle the resolutions?

https://gist.github.com/chrisronline/8cc094cd1876e895746d5e91db84be7c

Reply:
This working properly actually uncovers a UX issue in the UI where we don't really surface this well. We should really defer this to a separate PR IMO.